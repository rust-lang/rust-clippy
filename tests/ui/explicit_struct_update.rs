#![warn(clippy::explicit_struct_update)]

struct A {
    a: i32,
    b: i32,
    c: i32,
    d: i32,
}

struct B;

struct C {
    a: i32,
    b: i32,
}

fn main() {
    // should not lint, no explicit struct update
    let a = A { a: 1, b: 2, c: 3, d: 4 };

    let b = A {
        a: a.a,
        b: a.b,
        c: a.c,
        d: a.d,
    };
    //~^^^^^^explicit_struct_update

    let c = A {
        a: a.a,
        b: a.b,
        c: a.c,
        d: 5,
    };
    //~^^^^^^explicit_struct_update

    let d = A {
        a: a.a,
        b: 5,
        c: 6,
        d: 7,
    };
    //~^^^^^^explicit_struct_update

    // should not lint, we already have update syntax
    let e = A { ..a };

    // should not lint, we already have update syntax
    let f = A { a: a.a, b: a.b, ..a };

    // should not lint, multiple bases
    let g = A {
        a: a.a,
        b: d.b,
        c: d.c,
        d: 5,
    };

    // should not lint, no fields
    let h = B {};

    // should not lint, no explicit struct update
    let i = C { a: 1, b: 2 };

    // should not lint, fields filled from different type
    let j = A {
        a: i.a,
        b: i.b,
        c: 3,
        d: 4,
    };
}
