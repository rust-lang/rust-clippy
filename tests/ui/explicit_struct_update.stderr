error: you seem to be updating a struct field with the same field from another struct of the same type
  --> tests/ui/explicit_struct_update.rs:21:13
   |
LL |       let b = A {
   |  _____________^
LL | |         a: a.a,
LL | |         b: a.b,
LL | |         c: a.c,
LL | |         d: a.d,
LL | |     };
   | |_____^
   |
   = note: `-D clippy::explicit-struct-update` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::explicit_struct_update)]`
help: consider using struct update syntax instead
   |
LL ~     let b = A {
LL +         ..a
LL ~     };
   |

error: you seem to be updating a struct field with the same field from another struct of the same type
  --> tests/ui/explicit_struct_update.rs:29:13
   |
LL |       let c = A {
   |  _____________^
LL | |         a: a.a,
LL | |         b: a.b,
LL | |         c: a.c,
LL | |         d: 5,
LL | |     };
   | |_____^
   |
help: consider using struct update syntax instead
   |
LL ~     let c = A {
LL +         d: 5,
LL +         ..a
LL ~     };
   |

error: you seem to be updating a struct field with the same field from another struct of the same type
  --> tests/ui/explicit_struct_update.rs:37:13
   |
LL |       let d = A {
   |  _____________^
LL | |         a: a.a,
LL | |         b: 5,
LL | |         c: 6,
LL | |         d: 7,
LL | |     };
   | |_____^
   |
help: consider using struct update syntax instead
   |
LL ~     let d = A {
LL +         b: 5,
LL +         c: 6,
LL +         d: 7,
LL +         ..a
LL ~     };
   |

error: aborting due to 3 previous errors

